lfscs-meetings/resources/Minimum_Linux_Requirements_for_Executing_a_Minimal_Application.md

```diff
@@ -0,0 +1,455 @@
+
+### **Transforming ftrace Logs into Graphs**
+
+To make sense of the collected ftrace data, we first transform it into a 
+graphical representation. By using graph-tool in conjunction with ks-nav: a 
+tool that provides static kernel function call graphs, we generate visual 
+mappings of kernel execution paths.
+
+By limiting graph depth, we manage complexity, ensuring the readability of 
+results without omitting essential details. For example, a graph of the 
+`execve` system call limited to five levels may contain only a small fraction 
+of the complete execution path. Expanding this view reveals the intricate 
+nature of function calls: while a small graph may show just 369 connections, 
+the complete execution flow includes over 9000 edges, underscoring the 
+importance of structured analysis.
+
+### **Challenges with ftrace Logs and Kernel Interrupts**
+
+Although ftrace aims to capture kernel interactions initiated by `min.c`, 
+external factors introduce noise into the trace. One such complication is 
+CPU interrupts. Since interrupts do not have their own process identifiers 
+(PIDs), they borrow the PID of the process they interrupt. 
+This results in unexpected function calls appearing in the execution graph, 
+forming what we term “orphan nodes”: isolated points that do not fit neatly 
+into the program’s execution flow.
+
+### **Missing Data and the Role of Non-Traceable Functions**
+
+Tracing kernel execution is not always a complete process. Some functions are 
+explicitly marked as non-traceable using the `noinstr` attribute, preventing 
+them from appearing in ftrace logs. Additionally, compiler optimizations can 
+exclude function calls from tracing, creating gaps in the recorded execution 
+path.
+
+To assess these missing traces, we compare the kernel’s symbol table against 
+the list of traceable functions:
+
+```
+cat /proc/kallsyms | cut -d " " -f3 | sort > all
+sudo cat /sys/kernel/tracing/available_filter_functions | cut -d " " -f1 | sort > avail
+vimdiff all avail
+```
+
+This analysis reveals that numerous functions remain untraceable, 
+necessitating alternative methods to infer missing data.
+
+### **Filling the Gaps with Static Analysis**
+
+To compensate for missing function calls in our ftrace log, we leverage 
+`ks-nav`’s static analysis capabilities. By overlaying a static function call 
+graph onto the dynamically traced execution flow, we can infer missing 
+connections. The complete function of `graph-tool` enables us to merge 
+these datasets, visually distinguishing inferred connections.
+
+For example, integrating a static graph of `execve` with the ftrace-derived 
+graph introduces additional inferred edges:
+
+```
+./graph-tool complete -f1 executed.dot -f2 execve.dot -color green | xdot -
+```
+
+While this approach improves completeness, it is not foolproof... Some 
+indirect calls, such as those involving function pointers, remain unresolved.
+
```

Comment:
At this point it would be useful to show the initial graph image and how it was compensated by ks-nav